### PR TITLE
feat: segment eventing for LCM

### PR DESCRIPTION
**Pull request by brobro10000**
Adds segmenting events for learner credit management. 

# For all changes

- [ ] Ensure adequate tests are in place (or reviewed existing tests cover changes)

# Only if submitting a visual change

- [ ] Ensure to attach screenshots
- [ ] Ensure to have UX team confirm screenshots

### CODE REVIEW

File: src/components/learner-credit-management/cards/StackedCollapsible.jsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
+import PropTypes from 'prop-types';
```
Comment (brobro10000) [inline src/components/learner-credit-management/cards/StackedCollapsible.jsx:1]: Consolidated Collapsibles.jsx to avoid prop drilling

File: src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx [hunk 2]
```diff
@@ -41,9 +44,25 @@ const NewAssignmentModalButton = ({ course, children }) => {
 
   const pathToActivityTab = generatePath(routeMatch.path, { budgetId: subsidyAccessPolicyId, activeTabKey: 'activity' });
 
+  const handleOpenAssignmentModal = () => {
+    open();
+    sendEnterpriseTrackEvent(
+      enterpriseId,
+      EVENT_NAMES.LEARNER_CREDIT_MANAGEMENT.OPEN_ASSIGNMENT_MODAL_ASSIGN_COURSE,
+      {
+        isOpen: !isOpen,
```
Comment (adamstankiewicz) [inline src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx:53]: [curious] I'm wondering if the `isOpen` property is largely derived from the event name? For example, would the event name `EVENT_NAMES.LEARNER_CREDIT_MANAGEMENT.OPEN_ASSIGNMENT_MODAL_ASSIGN_COURSE` always indicate that the modal was opened?

Similar comment for the `handleCloseAssignmentModal` below, too, wrt to the `isOpen` property.

File: src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx [hunk 3]
```diff
@@ -84,32 +124,63 @@ const NewAssignmentModalButton = ({ course, children }) => {
           httpErrorStatus,
           httpErrorResponseData,
         } = err.customAttributes;
+        let locallyScopedErrorReason = 'system_error';
         if (httpErrorStatus === 422) {
           const responseData = JSON.parse(httpErrorResponseData);
           setCreateAssignmentsErrorReason(responseData[0].reason);
+          locallyScopedErrorReason = responseData[0].reason;
         } else {
           setCreateAssignmentsErrorReason('system_error');
```
Comment (adamstankiewicz) [inline src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx:133]: nit: I'm wondering if this line should rely on the `locallyScopedErrorReason` such that the string constant `system_error` doesn't need to be repeated twice in this function.

Might also consider renaming `locallyScopedErrorReason` to just `errorReason`, e.g.:

```jsx
let errorReason = 'system_error';
if (httpErrorStatus === 422) {
  const responseData = JSON.parse(httpErrorResponseData);
  errorReason = responseData[0].reason;
  setCreateAssignmentsErrorReason(errorReason);
} else {
  setCreateAssignmentsErrorReason(errorReason);
}
setAssignButtonState('error');
sendEnterpriseTrackEvent(...);
```

File: src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx [hunk 4]
```diff
@@ -84,32 +124,63 @@ const NewAssignmentModalButton = ({ course, children }) => {
...
       <FullscreenModal
         className="bg-light-200 text-left"
         title="Assign this course"
         isOpen={isOpen}
-        onClose={handleCloseAssignmentModal}
+        onClose={() => {
+          handleCloseAssignmentModal();
+          sendEnterpriseTrackEvent(
+            enterpriseId,
+            EVENT_NAMES.LEARNER_CREDIT_MANAGEMENT.CLOSE_ASSIGNMENT_MODAL_EXIT,
```
Comment (adamstankiewicz) [inline src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx:161]: [curious/clarification] Would this result in duplicate events regarding the modal closing? For example, this callback function would fire both `EVENT_NAMES.LEARNER_CREDIT_MANAGEMENT.CLOSE_ASSIGNMENT_MODAL_EXIT` and `EVENT_NAMES.LEARNER_CREDIT_MANAGEMENT.CLOSE_ASSIGNMENT_MODAL`.

File: src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx [hunk 5]
```diff
@@ -84,32 +124,63 @@ const NewAssignmentModalButton = ({ course, children }) => {
...
-            <Button variant="tertiary" onClick={close}>Cancel</Button>
+            <Button
+              variant="tertiary"
+              onClick={() => {
+                handleCloseAssignmentModal();
+                sendEnterpriseTrackEvent(
+                  enterpriseId,
+                  EVENT_NAMES.LEARNER_CREDIT_MANAGEMENT.CLOSE_ASSIGNMENT_MODAL_CANCEL,
+                  { assignButtonState },
+                );
```
Comment (adamstankiewicz) [inline src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx:179]: [curious/clarification] Would this result in duplicate events regarding the modal closing? For example, this callback function would fire both `EVENT_NAMES.LEARNER_CREDIT_MANAGEMENT.CLOSE_ASSIGNMENT_MODAL_CANCEL` and `EVENT_NAMES.LEARNER_CREDIT_MANAGEMENT.CLOSE_ASSIGNMENT_MODAL`.

File: src/components/learner-credit-management/data/tests/constants.test.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+/* eslint-disable react/jsx-filename-extension */
```
Comment (adamstankiewicz) [inline src/components/learner-credit-management/data/tests/constants.test.js:1]: [curious/clarification] Is this ESLint disable rule still relevant and/or could it be removed?

File: src/components/learner-credit-management/search/CatalogSearch.jsx [hunk 7]
```diff
@@ -7,17 +7,18 @@ import { SearchHeader } from '@edx/frontend-enterprise-catalog-search';
...
+} from '../data';
 
 const CatalogSearch = () => {
   const searchClient = algoliasearch(configuration.ALGOLIA.APP_ID, configuration.ALGOLIA.SEARCH_API_KEY);
   const { subsidyAccessPolicyId } = useBudgetId();
   const {
     data: subsidyAccessPolicy,
   } = useSubsidyAccessPolicy(subsidyAccessPolicyId);
-  const searchFilters = `enterprise_catalog_uuids:${subsidyAccessPolicy.catalogUuid} AND content_type:course`;
+  const searchFilters = `enterprise_catalog_uuids:${ENABLE_TESTING(subsidyAccessPolicy.catalogUuid)} AND content_type:course`;
```
Comment (adamstankiewicz) [inline src/components/learner-credit-management/search/CatalogSearch.jsx:20]: Note: The alternative here would be to ensure your local subsidy access policy points to a stage catalog via Django Admin without requiring any code changes related to `ENABLE_TESTING`. Or, the other alternative would be to run the local MFE against stage APIs (localhost.stage.edx.org:1991).

File: src/eventTracking.js [hunk 8]
```diff
@@ -94,6 +94,22 @@ export const SUBSCRIPTION_EVENTS = {
...
+  CLOSE_ASSIGNMENT_MODAL_ASSIGNED: `${LEARNER_CREDIT_MANAGEMENT_PREFIX}.assignment-modal.close-modal-assigned.clicked`,
+  CLOSE_ASSIGNMENT_MODAL_CANCEL: `${LEARNER_CREDIT_MANAGEMENT_PREFIX}.assignment-modal.close-modal-cancel.clicked`,
+  CLOSE_ASSIGNMENT_MODAL_EXIT: `${LEARNER_CREDIT_MANAGEMENT_PREFIX}.assignment-modal.close-modal-exit.clicked`,
+  CLOSE_ASSIGNMENT_MODAL_ASSIGNMENT_ALLOCATION_ERROR: `${LEARNER_CREDIT_MANAGEMENT_PREFIX}.assignment-modal.exit-assignment-allocation-modal.clicked`,
+  // Successful allocation
+  ASSIGNMENT_ALLOCATED_LEARNER_COUNT: `${LEARNER_CREDIT_MANAGEMENT_PREFIX}.assignment-modal.assignment-allocation.assigned`,
+  // Input Validation and API error allocation
+  EMAIL_ADDRESS_VALIDATION: `${LEARNER_CREDIT_MANAGEMENT_PREFIX}.assignment-modal.email-validation.changed`,
+  ASSIGNMENT_ALLOCATION_ERROR: `${LEARNER_CREDIT_MANAGEMENT_PREFIX}.assignment-modal.assignment-allocation.errored`,
 };
```
Comment (adamstankiewicz) [inline src/eventTracking.js:113]: nit: Something I just realized... these and many our existing event names have mixed usage of `_` and `-` to separate works. I wonder if we should ensure at least in these `LEARNER_CREDIT_MANAGEMENT_EVENTS` that we're only using either `_` OR `-`. That way, when we go to query against these events in a Snowflake dashboard, it won't be as confusing since the event name would contain only `_` or only `-`.

I.e., The `LEARNER_CREDIT_MANAGEMENT_PREFIX` relies on `_` but the concatenated event specific strings rely on `-`.

File: src/components/learner-credit-management/cards/CourseCard.jsx [hunk 9]
```diff
@@ -39,4 +15,4 @@ CourseCard.propTypes = {
   original: PropTypes.shape().isRequired, // pass-thru prop to `BaseCourseCard`
 };
 
-export default CourseCard;
+export default (CourseCard);
```
[the remark this answers is not in the document]
Comment (brobro10000) [reply src/components/learner-credit-management/cards/CourseCard.jsx:18]: lol a previous attempt to `connect(mapStatetoProps)(CourseCard)`

File: src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx [hunk 10]
```diff
@@ -57,6 +76,25 @@ const NewAssignmentModalButton = ({ course, children }) => {
...
 
+  const onSuccessEnterpriseTrackEvents = () => {
+    sendEnterpriseTrackEvent(
+      enterpriseId,
+      EVENT_NAMES.LEARNER_CREDIT_MANAGEMENT.CLOSE_ASSIGNMENT_MODAL_ASSIGNED,
+      {
+        isOpen: !isOpen,
+        assignButtonState,
+      },
+    );
```
[the remark this answers is not in the document]
Comment (brobro10000) [reply src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx:87]: There was an additional track event in `handleCloseAssignmentModal` which I forgot to refactor out actually 👍🏽  great catch on that. (Opted for distinct events to handle exit and close instead of generalizing the solution).

I like the idea of consolidating the events and will give it a refactor now. 💯

File: src/eventTracking.js (hunk 8 above)
Comment (brobro10000) [reply src/eventTracking.js:113]: opting for the `_` over `-` since according to the previous convention of content highlights, we are also using underscore, along with all the prefixes

File: src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx (hunk 5 above)
Comment (brobro10000) [reply src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx:177]: Resolved by removing duplicate track event from `handleCloseAssignmentModal()`.

File: src/components/learner-credit-management/cards/data/utils.js [hunk 11]
```diff
@@ -53,7 +53,7 @@ export const isEmailAddressesInputValueValid = ({
   const hasEnoughBalanceForAssigment = remainingBalanceAfterAssignment >= 0;
 
   const invalidEmails = learnerEmails.filter((email) => !isEmail(email));
-  const duplicateEmails = learnerEmails.filter((email, index) => learnerEmails.indexOf(email) !== index);
+  const duplicateEmails = learnerEmails.filter((email, index) => learnerEmails.indexOf(email.toLowerCase()) !== index);
```
Comment (brobro10000) [inline src/components/learner-credit-management/cards/data/utils.js:56]: handles verification edge case where the user may enter an uppercase and lower case duplicate email. (TEST@gmail.com, test@gmail.com).
Verification at this point in the code results in the validation error retaining the casing of the error, relaying to the user to remove either the uppercase or lowercase value email address.

File: src/components/learner-credit-management/data/tests/constants.test.js (hunk 6 above)
Comment (brobro10000) [reply src/components/learner-credit-management/data/tests/constants.test.js:1]: I guess it got autogenerated from my IDEs auto-linting feature. Removed, and the warning doesn't seem to be replicable 🤷🏽

File: src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx (hunk 2 above)
Comment (brobro10000) [reply src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx:53]: It might be worth removing the OPEN/CLOSE prefixes on segments event related to this and have the event metadata speak for itself whether it isOpen or !isOpen 😺

File: src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx [hunk 12]
```diff
@@ -57,6 +76,25 @@ const NewAssignmentModalButton = ({ course, children }) => {
...
+    sendEnterpriseTrackEvent(
+      enterpriseId,
+      EVENT_NAMES.LEARNER_CREDIT_MANAGEMENT.ASSIGNMENT_ALLOCATED_LEARNER_COUNT,
+      {
+        allocatedLearners: learnerEmails.length,
+        courseUUID: course.uuid,
```
[the remark this answers is not in the document]
Comment (brobro10000) [reply src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx]: I was able to successfully destructure out the relevant data. 
Screenshot 2023-11-26 at 1 39 54 PM

File: src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx (hunk 4 above)
Comment (brobro10000) [reply src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx]: Resolved in an earlier comment 👍🏽

File: src/components/learner-credit-management/cards/data/utils.js (hunk 11 above)
Comment (brobro10000) [reply src/components/learner-credit-management/cards/data/utils.js]: Chose to resolve it in a separate PR, https://github.com/openedx/frontend-app-admin-portal/pull/1101

File: src/components/learner-credit-management/BudgetDetailAssignments.jsx [hunk 13]
```diff
@@ -34,7 +36,6 @@ const BudgetDetailAssignments = ({
       <AssignMoreCoursesEmptyStateMinimal />
     );
   }
-
```
[the remark this answers is not in the document]
Comment (brobro10000) [reply src/components/learner-credit-management/BudgetDetailAssignments.jsx:37]: No particular pattern, but most likely due to a previous console log right before the return that got cut out and the space was not added back in.

I'll remember to stick to the space before return convention (which I also generally prefer).

File: src/components/learner-credit-management/cards/AssignmentModalContent.jsx [hunk 14]
```diff
@@ -11,18 +11,20 @@ import {
...
 } from '@edx/paragon';
+import { sendEnterpriseTrackEvent } from '@edx/frontend-enterprise-utils';
 
+import { connect } from 'react-redux';
 import BaseCourseCard from './BaseCourseCard';
 import { formatPrice, useBudgetId, useSubsidyAccessPolicy } from '../data';
-import { ImpactOnYourLearnerCreditBudget, ManagingThisAssignment, NextStepsForAssignedLearners } from './Collapsibles';
 import AssignmentModalSummary from './AssignmentModalSummary';
 import { EMAIL_ADDRESSES_INPUT_VALUE_DEBOUNCE_DELAY, isEmailAddressesInputValueValid } from './data';
+import CollapsibleStack from './StackedCollapsible';
```
[the remark this answers is not in the document]
Comment (brobro10000) [reply src/components/learner-credit-management/cards/AssignmentModalContent.jsx:21]: Right click - rename refactor ftw :)

File: src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx [hunk 15]
```diff
@@ -84,32 +117,73 @@ const NewAssignmentModalButton = ({ course, children }) => {
...
         footerNode={(
           <ActionRow>
-            <Button variant="tertiary" as={Hyperlink} destination="https://edx.org" target="_blank">
+            <Button
+              variant="tertiary"
+              as={Hyperlink}
+              onClick={() => sendEnterpriseTrackEvent(
+                enterpriseId,
+                EVENT_NAMES.LEARNER_CREDIT_MANAGEMENT.ASSIGNMENT_MODAL_HELP_CENTER,
+                {},
```
[the remark this answers is not in the document]
Comment (brobro10000) [reply src/components/learner-credit-management/cards/NewAssignmentModalButton.jsx:166]: That I did not know. Will refactor those empty objects away 👍🏽

File: src/components/learner-credit-management/data/hooks/tests/useBudgetContentAssignments.test.js [hunk 16]
```diff
@@ -91,10 +102,13 @@ describe('useBudgetContentAssignments', () => {
       ],
       hasSearchParam: false,
     },
-  ])('handles assignment details filter with search query parameter (%s)', async ({ filters, hasSearchParam }) => {
+  ])('handles assignment details filter with search query parameter (%s)', async ({
+    filters, sortBy, hasSearchParam,
```
[the remark this answers is not in the document]
Comment (brobro10000) [reply src/components/learner-credit-management/data/hooks/tests/useBudgetContentAssignments.test.js:106]: Stray remnant variable of a test refactor. Repurposed an existing test instead and it got left behind 👍🏽

File: src/eventTracking.js [hunk 17]
```diff
@@ -93,7 +93,26 @@ export const SUBSCRIPTION_EVENTS = {
...
 
 export const LEARNER_CREDIT_MANAGEMENT_EVENTS = {
-  TAB_CHANGED: `${LEARNER_CREDIT_MANAGEMENT_PREFIX}.budget-detail.tab.changed`,
+  TAB_CHANGED: `${LEARNER_CREDIT_MANAGEMENT_PREFIX}.budget_detail.tab.changed`,
+  // activity tab
+  BUDGET_DETAILS_DATATABLE_SORT_BY_OR_FILTER: `${LEARNER_CREDIT_MANAGEMENT_PREFIX}.budget_details_datatable.sort_by_or_filter.changed`,
```
[the remark this answers is not in the document]
Comment (brobro10000) [reply src/eventTracking.js:98]: Defined specific abstractions for future iterations on the feature.
Screenshot 2023-11-28 at 11 53 41 AM

File: src/components/learner-credit-management/tests/NoBudgetActivityEmptyState.test.jsx [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+    });
+    useRouteMatch.mockReturnValue({
+      path: '/',
+    });
+    renderWithRouter(<NoBudgetActivityEmptyStateWrapper />);
+
+    userEvent.click(screen.getByText('Get started'));
+    await waitFor(() => expect(sendEnterpriseTrackEvent).toHaveBeenCalledTimes(1));
+  });
+});
```
[the remark this answers is not in the document]
Comment (brobro10000) [reply src/components/learner-credit-management/tests/NoBudgetActivityEmptyState.test.jsx]: refactored test away and used the BudgetDetailPage test file instead 👍🏽

File: src/components/learner-credit-management/AssignmentsTableRefreshAction.jsx [hunk 19]
```diff
@@ -1,11 +1,37 @@
...
+      page: dataTableState.pageIndex + 1, // `DataTable` uses zeo-indexed array
+      pageSize: dataTableState.pageSize,
+    };
+    applyFiltersToOptions(dataTableState.filters, options);
+    applySortByToOptions(dataTableState.sortBy, options);
+    const trackEventMetadata = {
+      filters: {
+        learnerState: options?.learnerState || null,
+        search: options?.search || null,
+      },
```
Comment (adamstankiewicz) [inline src/components/learner-credit-management/AssignmentsTableRefreshAction.jsx:25]: nit/curious: why is `options` handled as if it may be undefined/null here whereas below the `?` is not included?

File: src/components/learner-credit-management/data/hooks/tests/useBudgetContentAssignments.test.js [hunk 20]
```diff
@@ -279,5 +292,66 @@ describe('useBudgetContentAssignments', () => {
         ordering: orderingQueryParam,
       },
     );
+
+    // expect(sendEnterpriseTrackEvent).toHaveBeenCalledTimes(1);
```
Comment (adamstankiewicz) [inline src/components/learner-credit-management/data/hooks/tests/useBudgetContentAssignments.test.js:296]: Should this commented out line be un-commented or removed if no longer necessary?

File: src/components/learner-credit-management/data/hooks/tests/useBudgetContentAssignments.test.js [hunk 21]
```diff
@@ -279,5 +292,66 @@ describe('useBudgetContentAssignments', () => {
...
+      pageSize: 10,
+      sortBy: [initialSortByMetadata],
+    });
+
+    await waitForNextUpdate();
+
+    expect(sendEnterpriseTrackEvent).not.toHaveBeenCalled();
+
+    // Perform second render after the currentArgsRef.current has been hydrated
+    renderHook(() => useBudgetContentAssignments(mockUseBudgetContentAssignmentsData));
```
Comment (adamstankiewicz) [inline src/components/learner-credit-management/data/hooks/tests/useBudgetContentAssignments.test.js:344]: [suggestion] I believe the `renderHook` function returns a `rerender` function ([docs](https://react-hooks-testing-library.com/reference/api#rerender)) you could call in place of calling `renderHook` itself.

File: src/components/learner-credit-management/data/hooks/useBudgetContentAssignments.js [hunk 22]
```diff
@@ -76,15 +81,45 @@ const useBudgetContentAssignments = ({
...
+
+      const argsCopy = {
+        pageIndex: args?.pageIndex,
+        pageSize: args?.pageSize,
+        filters: args?.filters,
+        sortBy: args?.sortBy,
+      };
+      /* This logic in conjunction with useRef is being used to prevent track events
+      from being called when the page re-renders without the specifically selected
+      arguments (argCopy) being changed */
```
Comment (adamstankiewicz) [inline src/components/learner-credit-management/data/hooks/useBudgetContentAssignments.js:93]: nit: might recommend putting the comment above `argsCopy` as I personally would first have a question about why `argsCopy` is created before seeing how its used below.

File: src/components/learner-credit-management/AssignmentsTableRefreshAction.jsx (hunk 19 above)
Comment (brobro10000) [reply src/components/learner-credit-management/AssignmentsTableRefreshAction.jsx:25]: I went ahead and continued the behavior for the rest of the attributes, and mirrored the same definition in the trackEventMetadata object in `useBudgetContentAssignments`

File: src/components/learner-credit-management/data/hooks/tests/useBudgetContentAssignments.test.js (hunk 20 above)
Comment (brobro10000) [reply src/components/learner-credit-management/data/hooks/tests/useBudgetContentAssignments.test.js:296]: Removed 👍🏽

File: src/components/learner-credit-management/data/hooks/useBudgetContentAssignments.js (hunk 22 above)
Comment (brobro10000) [reply src/components/learner-credit-management/data/hooks/useBudgetContentAssignments.js:93]: Done 👍🏽

File: src/components/learner-credit-management/data/hooks/tests/useBudgetContentAssignments.test.js (hunk 21 above)
Comment (brobro10000) [reply src/components/learner-credit-management/data/hooks/tests/useBudgetContentAssignments.test.js:344]: Worked as expected, but still requires us to mock a second `listContentAssignments` and refetch the `fetchContentAssignments` 👍🏽

**Review by adamstankiewicz — APPROVED** [review]
LGTM!
